### PR TITLE
[MRG] ENH Control Bayesian optimizer seed

### LINKED ISSUE #1294: [FEAT] Request: Add ability to control random seeding in Bayesian optimization
Currently there is no way to fix/control the random seed during Bayesian optimization. The `bayes_opt` function actually has a parameter to control this:
https://github.com/jonescompneurolab/hnn-core/blob/dd676043bf39cf8c089bb5cb72f9316d838096d7/hnn_core/optimization/general_optimization.py#L445

So exposing this would be a matter of passing it to `optim.fit(..., seed=123)` where the seed is included as an entry in `obj_fun_kwargs`, and then internally gets passed to `bayes_opt()`.

### PR DESCRIPTION
**Pull request by floze-the-genius**
## Summary

- Forward the seed passed to Optimizer.fit into Bayesian optimization.
- Use that random state for candidate sampling as well as the Gaussian process.
- Document Bayesian seed support and cover repeated and distinct seeds with a regression test.

Fixes #1294

## Validation

- pytest -q hnn_core/tests/test_general_optimization.py (34 passed)
- make format-check lint spell

### CODE REVIEW

**Comment by ntolley** [comment]
Thanks for the PR @floze-the-genius !! 

Unfortunately it seems like github's servers are down right now preventing me from seeing the code changes but I will definitely keep an eye out for when they return

File: hnn_core/optimization/general_optimization.py [hunk 1]
```diff
@@ -480,6 +480,7 @@ def _obj_func(predicted_params):
         cons=constraints,
         acquisition=expected_improvement,
         maxfun=max_iter,
+        random_state=obj_fun_kwargs.get("seed"),
```
Comment (ntolley) [inline hnn_core/optimization/general_optimization.py:483]:
```suggestion
        random_state=obj_fun_kwargs.get("seed", 123),
```
I think we will actually have this be a default seed

File: hnn_core/externals/bayesopt.py
Comment (ntolley) [inline hnn_core/externals/bayesopt.py:49]:
```suggestion
def bayes_opt(func, x0, cons, acquisition, maxfun=200, debug=False, random_state=123):
```
Related to the change below, we will fix the default seed here

File: hnn_core/externals/bayesopt.py [hunk 2]
```diff
@@ -83,13 +84,14 @@ def bayes_opt(func, x0, cons, acquisition, maxfun=200, debug=False, random_state
     best_x = X[np.argmin(y)]
     best_f = y[np.argmin(y)]
     gp = gaussian_process.GaussianProcessRegressor(random_state=random_state)
+    rng = check_random_state(random_state)
```
Comment (ntolley) [inline hnn_core/externals/bayesopt.py:87]:
```suggestion
    rng = np.random.default_rng(random_state)
```

File: hnn_core/externals/bayesopt.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 
 import warnings
 from sklearn import gaussian_process
+from sklearn.utils import check_random_state
```
Comment (ntolley) [inline hnn_core/externals/bayesopt.py:16]:
```suggestion
```

**Review by ntolley — CHANGES_REQUESTED** [review]
Awesome work @floze-the-genius!!

I only have a few minor suggestions but otherwise this PR is good to go 🚀

**Comment by asoplata** [comment]
I reviewed along with Nick, and I agree with all the suggested changes by Nick. After these changes are made, this PR should be good to merge. Thanks Floze!
